### PR TITLE
limit time and memory consumption

### PR DESCRIPTION
**Pull request by sonichi**
This PR addresses time and memory limit during training. Try to prevent OOM error or timeout.
- Prevent lrl2 from training for too long by monitoring the time and raising TimeoutError
- For lrl2, use the resource limit to raise MemoryError
- For lgbm, xgb (>=1.3), catboost (>=0.26), use callback to check time and memory
- For rf, extra_trees, xgb (<1.3), catboost(<0.26), estimate the memory usage and modify n_estimators when needed
- optimize the memory usage. We will not save models during tuning unless model_history=True. That means best_model_per_estimator will return None for non best estimators if model_history=False. Maybe we can remove model_history and add save_best_model_per_estimator in future.
- fixed bugs about retrain_final when max_iter=1, verbosity, best_config_train_time when retraining.
- xgboost requires libomp version no higher than 11.1.0 on macOS. Using a workaround in CI. 

Limitation: memory and time limit for lrl2 are unavailable in Windows (`import resource` fails and signal doesn't have SIGALRM); lrl1 can't be limited this way (error will be raised if we limit the memory and the signal doesn't interrupt the training); callback can't prevent OOM if training one iteration causes OOM (a workaround is to use ray even when n_concurrent_trials = 1 which will run each training in separate processes). Help in addressing these limitations is appreciated.

### CODE REVIEW

File: flaml/model.py [hunk 1]
```diff
@@ -2,20 +2,67 @@
...
+            main_thread = True
+        except ValueError:
+            pass
+    try:
+        yield
+    finally:
+        if main_thread:
+            signal.alarm(0)
+        if memory_limit > 0:
+            resource.setrlimit(resource.RLIMIT_AS, (soft, hard))
```
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/model.py:65]: recover the previous limit.

File: flaml/model.py [hunk 2]
```diff
@@ -2,20 +2,67 @@
...
+
+def TimeoutHandler(sig, frame):
+    raise TimeoutError(sig, frame)
+
+
+@contextmanager
+def limit_resource(memory_limit, time_limit):
+    if memory_limit > 0:
+        soft, hard = resource.getrlimit(resource.RLIMIT_AS)
+        if soft < 0 or memory_limit < soft:
```
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/model.py:49]: why? We are only modifying soft.

[the remark this answers is not in the document]
Comment (qingyun-wu) [reply flaml/model.py:49]: is soft allowed to exceed hard?

[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/model.py]: modified. Is it OK now?

**Review by qingyun-wu — APPROVED** [review]

**Pull request merged**